src/jace/jax/translation_cache.py

```diff
@@ -0,0 +1,297 @@
+
+
+def clear_translation_cache() -> None:
+    """Clear all caches associated to translation."""
+    _TRANSLATION_CACHES.clear()
+
+
+def get_cache(
+    stage: stages.Stage,
+) -> TranslationCache:
+    """Returns the cache that is used for `stage`."""
+    # The caches are per stage and not per instance basis
+    tstage = type(stage)
+    if tstage not in _TRANSLATION_CACHES:
+        _TRANSLATION_CACHES[tstage] = TranslationCache(size=_DEF_CACHE_SIZE)
+    return _TRANSLATION_CACHES[tstage]
+
+
+@dataclass(init=True, eq=True, frozen=True)
+class _AbstarctCallArgument:
+    """Class to represent one argument to the call in an abstract way.
+
+    It is used as part of the key in the cache.
+    It represents the structure of the argument, i.e. its shape, type and so on, but nots its value.
+    To construct it you should use the `from_value()` class function which interfere the characteristics from a value.
+    """
+
+    shape: tuple[int, ...]
+    dtype: dace.typeclass
+    strides: tuple[int, ...] | None
+    storage: dace.StorageType
+
+    @classmethod
+    def from_value(
+        cls,
+        val: Any,
+    ) -> _AbstarctCallArgument:
+        """Construct an `_AbstarctCallArgument` from a value.
+
+        Todo:
+            Improve, such that NumPy arrays are on CPU, CuPy on GPU and so on.
+            This function also probably fails for scalars.
+        """
+        if not util.is_fully_addressable(val):
+            raise NotImplementedError("Distributed arrays are not addressed yet.")
+        if isinstance(val, jax_core.Literal):
+            raise TypeError("Jax Literals are not supported as cache keys.")
+
+        if util.is_array(val):
+            if util.is_jax_array(val):
+                val = val.__array__(copy=False)
+            shape = val.shape
+            dtype = util.translate_dtype(val.dtype)
+            strides = getattr(val, "strides", None)
+            # TODO(phimuell): is `CPU_Heap` always okay? There would also be `CPU_Pinned`.
+            storage = (
+                dace.StorageType.GPU_Global if util.is_on_device(val) else dace.StorageType.CPU_Heap
+            )
+
+            return cls(shape=shape, dtype=dtype, strides=strides, storage=storage)
+
+        if util.is_scalar(val):
+            shape = ()
+            dtype = util.translate_dtype(type(val))
+            strides = None
+            # Lets pretend that scalars are always on the CPU, which is a fair assumption.
+            storage = dace.StorageType.CPU_Heap
+
+            return cls(shape=shape, dtype=dtype, strides=strides, storage=storage)
+
+        raise TypeError(f"Can not make 'an abstract description from '{type(val).__name__}'.")
+
+
+@runtime_checkable
+class _ConcreteCallArgument(Protocol):
+    """Type for encoding a concrete arguments in the cache."""
+
+    @abstractmethod
+    def __hash__(self) -> int:
+        pass
+
+    @abstractmethod
+    def __eq__(self, other: Any) -> bool:
+        pass
```

Comment:
This protocol is very generic and doesn't define any requirement linked to concrete call arguments so I'd recommend to use a generic name (e.g. `Comparable`) and consider moving it to a more general module to make it available in other parts of Jace.

Reply:
I deleted it and uses `Hashable`, however, I think it is less expressive.